add parallel-execution-framework.md (#146)
pingcap/tidb-dev-guide: src/understand-tidb/execution-framework.md

> # Execution Framework

Author: Maybe we should use `Parallelism Execution` instead of `Execution Framework`.

Reviewer: This section is named by @XuHuaiyu , what do you want to express at this section?

Reviewer 2: 1. We can rename it as `Parallel Execution Framework`.
2. We'd better add some examples or figures to make the description easier to be undertood. e.g. The difference of the implementation between intra-parallelism and inter-parallelism.

In order to make full use of the multi-core ability of modern hardware, most popular DBMS have implemented the ability of parallel execution SQL execution engine.

> There are three common parallel implementations:
> 1. Intra operator parallelism.
> 2. Exchange operator.
> 3. Morsel-Driven parallelism.

Reviewer: ```suggestion
There are three common parallel implementations: intra-operator parallelism, exchange operator, and Morsel-Driven parallelism.
```

As you have just three sections for each.

## Intra-operator Parallelism

> In intra-operator parallelism, multiple goroutines will be created inside the operator for parallel processing. The creation, termination and synchronization of the goroutines are handled by the operator itself.For example, in HashJoin, after the hashtable on the build side is built, multiple goroutines will lookup the hashtable at the same time to speed up the execution of HashJoin.And the parallel implementation of each operator is different. Hashagg is divided into two stages. Each stage will start multiple goroutines.It's still the traditional Volcano-model, and the `Next()` is still called by a single thread. The parallelism only occurs in the internal processing of the operator.

Reviewer: ```suggestion
In intra-operator parallelism, multiple goroutines will be created inside the operator for parallel processing. The creation, termination and synchronization of the goroutines are handled by the operator itself.

For example, in HashJoin, after the hashtable on the build side is built, multiple goroutines will lookup the hashtable at the same time to speed up the execution of HashJoin. 

The parallel implementation of operators is various. HashAgg is divided into two stages. Each stage will start multiple goroutines. It still follows the traditional Volcano-model, and the `Next()` method is still called by a single thread. The parallelism only occurs in the internal processing of the operator.
```

> The degree of parallelism(DOP) can be controlled by the session variable. For example, in TiDB `tidb_executor_Concurrency` is 5 by default. it means HashJoin will create 5 goroutines to probe HashTable when it's opened.You can also control the parallelism of a specific operator by setting the session variable, such as `tidb_hash_join_concurrency`.

Reviewer: ```suggestion
The degree of parallelism (DOP) can be controlled by the session variable. For example, `tidb_executor_concurrency` is 5 by default. It means HashJoin will create five goroutines to probe HashTable when it's opened. You can also control the parallelism of a specific operator by setting the session variable, such as `tidb_hash_join_concurrency`.
```

## Exchange Operator

Intra operator parallelism is the most intuitive way to implement parallelism, but the implementation is more complex, because every operator needs to implement parallelism independently.Moreover, too many threads will increase the scheduling overhead. However, the use of goroutines can alleviate this problem.

> The most traditional way to implement parallelism is to use the exchange operator. By encapsulating the parallel logic into the exchange operator, all other operators only need to consider the single thread implementation.The exchange operator is responsible for data exchange between different threads. The internal implementation of exchange is divided into two parts: sender and receiver.And data is transferred from sender to receiver in different ways, such as hash partition, random distribution or sort merge.

Reviewer: ```suggestion
A more traditional way to implement parallelism is to use the exchange operator. By encapsulating the parallel logic into the exchange operator, all other operators only need to consider the single thread implementation.

The exchange operator is responsible for data exchange between different threads. The internal implementation of exchange is divided into two parts: sender and receiver. And data is transferred from sender to receiver in different ways, such as hash partition, random distribution or sort merge.
```

> There are usually two ways to control DOP:

Reviewer: ```suggestion
There are usually two ways to control DOP:

```

1. Users can use hints to specify DOP explicitly.
2. The optimizer can choose DOP according to the table size of the scan operation automatically.

> This approach requires the optimizer to generate parallel plans. Generally, the generation of plans is divided into two stages. The first stage generates serial plans, and the second stage generates its corresponding parallel plans. The second stage is mainly responsible for inserting the exchange operator into the plan tree at the appropriate position. Both heuristic rules and costs model can be used to get the optimal parallel plan

Reviewer: ```suggestion
The latter approach requires the optimizer to generate parallel plans. Generally, plan generation is divided into two stages. The first stage generates serial plans, and the second stage generates its corresponding parallel plans. The second stage is mainly responsible for inserting the exchange operator into the plan tree at the appropriate position. Both heuristic rules and costs model can be used to get the optimal parallel plan.
```
